### ISSUE TITLE
Support modified OpenVPN exe

### ISSUE THREAD

**Issue by lgxz** [Tunnelblick/Tunnelblick#599]
**Describe the bug**
This is a feature request, not a bug.

**Expected behavior**
1. VPN details -> Configurations -> OpenVPN Version:
2. Instead of selecting from the dropdown list, I can choose a OpenVPN version from the file system.

Tunnelblick will warn me that it was tampered if I just copy the modified OpenVPN exe to /Applications/Tunnelblick.app/Contents/Resources/openvpn/openvpn-2.4.8-openssl-1.1.1d/openvpn.

**Comment by jkbullard** [comment]
@lgxz - Why? What are the modifications that you want to make to OpenVPN?

**Comment by lgxz** [comment]
> @lgxz - Why? What are the modifications that you want to make to OpenVPN?

To support the OTP auth.  Our OpenVPN server requires username and password (which is the account password appended by 6 digits OTP code).

I modified openvpn/misc.c:get_user_pass_cr function to convert "password:otp_secret" to "passwordOTP_CODE":

```
+static void otp_conv_passwd(char *passwd)
+{
+    char *p;
+    unsigned char sbytes[10];
+    int plen = strlen(passwd);
+
+    p = passwd + plen - 17;
+    if (plen > 17 && *p == ':' && b32decode(p+1, sbytes))
+    {
+        totp(sbytes, p);
+    }
+}
+
+
 /*
  * Get and store a username/password
  */
@@ -1139,6 +1195,8 @@
     msg(M_INFO, "GET_USER_PASS %s u='%s' p='%s'", prefix, up->username, up->password);
 #endif
 
+    otp_conv_passwd(up->password);
+
     gc_free(&gc);
 
     return true;
```

**Comment by lgxz** [comment]
> @lgxz - Why? What are the modifications that you want to make to OpenVPN?

I had patched Tunnelblick to support the OTP auth previously. But, the problem is that I can't upgrade to a new version of Tunnelblick. 

Yesterday I had the new idea. I can patch the openvpn instead of Tunnelblick. So, for every new version of Tunnelblick, I just copy the patched openvpn exe to /Applications/Tunnelblick.app/.../openvpn.

**Comment by jkbullard** [comment]
You may now use your own OpenVPN binaries with Tunnelblick. See [Using Custom OpenVPN Binaries](https://tunnelblick.net/cUsingCustomOpenVPNBinaries.html).

**Comment by lgxz** [comment]
> @lgxz - You may now use your own OpenVPN binaries with Tunnelblick. See [Using Custom OpenVPN Binaries](https://tunnelblick.net/cUsingCustomOpenVPNBinaries.html).

A lot of changes...  Thanks very much!

But I failed to get it works.  It silently fell back to openvpn 2.3.18 despite there was a log message about external openvpn.

`
2020-02-19 21:00:55.649024 *Tunnelblick: openvpnstart start test.tblk 54676 769 0 1 0 1098032 -ptADGNWradsgnw 2.4.8-openssl-1.1.1d_TestEXTERNAL
2020-02-19 21:00:56.274399 *Tunnelblick: 

Could not start OpenVPN (openvpnstart returned with status #251)

Contents of the openvpnstart log:
openvpnstart log:
     OpenVPN returned with status 1, errno = 0:
          Undefined error: 0
     
     Command used to start OpenVPN (one argument per displayed line):
          /Applications/Tunnelblick.app/Contents/Resources/openvpn/openvpn-2.3.18-libressl-2.7.1/openvpn
          --daemon
......
     Contents of the OpenVPN log:
     
          Options error: Unrecognized option or missing parameter(s) in /Library/Application Support/Tunnelblick/Users/lgx/test.tblk/Contents/Resources/config.ovpn:18: compress (2.3.18)
          Use --help for more information.
          
`

**Comment by jkbullard** [comment]
1. Please look for error messages in the [Console Log](https://tunnelblick.net/cConsoleLog.html).

 2. Please post the output from the following command:
<pre>
     ls -lR "/Library/Application Support/Tunnelblick/Openvpn"
</pre>

**Comment by lgxz** [comment]
1. Console log:

Feb 20 15:21:40 P tunnelblickd[24951]: Status = 251 from tunnelblick-helper command 'start ks.tblk 64361 769 0 1 0 1098032 -ptADGNWradsgnw 2.4.8-openssl-1.1.1d_TestEXTERNAL'
Feb 20 15:21:40 P Tunnelblick[23904]: tunnelblickd status from start: 251
Feb 20 15:21:48 P Tunnelblick[23904]: The OpenVPN log contains the following message: 
	     • is not implemented by the version of OpenVPN which is being used for this configuration. It may be a new option that is not implemented in an old version of OpenVPN, or an old option that has been removed in a new version of OpenVPN. You can choose what version of OpenVPN to use with this configuration in the "Settings" tab of the "Configurations" panel of Tunnelblick's "VPN Details" window.
	See the VPN log in the "Log" tab of the "Configurations" panel of Tunnelblick's "VPN Details" window for details.


2. ls -lR "/Library/Application Support/Tunnelblick/Openvpn"

% ls -lR "/Library/Application Support/Tunnelblick/Openvpn"
total 0
drwxr-xr-x  4 root  wheel  128 Feb 19 12:43 openvpn-2.4.8-openssl-1.1.1d_Test

/Library/Application Support/Tunnelblick/Openvpn/openvpn-2.4.8-openssl-1.1.1d_Test:
total 8264
-rwxr-xr-x@ 1 root  wheel  3162868 Feb 19 12:56 openvpn
-rwxr--r--@ 1 root  wheel    36720 Jan 30 13:33 openvpn-down-root.so

**Comment by lgxz** [comment]
> 1. Please look for error messages in the [Console Log](https://tunnelblick.net/cConsoleLog.html).
> 2. Please post the output from the following command:
> 
>      ls -lR "/Library/Application Support/Tunnelblick/Openvpn"

In addition, there was a popup message said that 'openvpn' was downloaded from network and was not trusted when the Tunnelblick started firstly.  I allowed the execution of 'openvpn' in the MacOS System Preferences->Security and Privacy.

**Comment by jkbullard** [comment]
There's bug in Tunnelblick: If the name of the "openvn-n.n.n-openssl-n.n.n…" folder is more than 31 characters long, Tunnelblick ignores it! (This was because the folders created by Tunnelblick always had a name shorter than 32 characters long. That limit should have been removed when we added this feature, because the user controls the folder name, so it may be more than 32 characters long.

This will be fixed in the next beta release of Tunnelblick.

Workaround: The name of your folder is "openvpn-2.4.8-openssl-1.1.1d_Test", which is 33 characters long! Remove two characters and it should work: "openvpn-2.4.8-openssl-1.1.1d_Te", for example.

(This happened because the folders created by Tunnelblick always had a name shorter than 32 characters long. That limit should have been removed when we added this feature, because the user controls the folder name, so it may be more than 32 characters long, but that wasn't done.)

**Comment by lgxz** [comment]
Wow... There are too many security checks in the Tunnelblick source code. Is the 32-chars restriction used to prevent some attack method?

Now, the 'openvpn' was called, but there's still a problem (In the Tunnelblick log window):


```
Could not start OpenVPN (openvpnstart returned with status #208)

Contents of the openvpnstart log:
openvpnstart log:
     Path /Library/Application Support/Tunnelblick/Openvpn/openvpn-2.4.8-openssl-1.1.1d_Te/openvpn may not be run as root

```

No more meaningful log in the Console log.

**Comment by lgxz** [comment]
It seems that there are additional checks in tunnelblick-helper.m : exitIfPathShouldNotBeRunAsRoot?

**Comment by lgxz** [comment]
I thought that it's a simple feature request, not expect it would involve so many problems ...

**Comment by jkbullard** [comment]
Yes, such limits (on things known to be limited in size) are meant to prevent out-of-memory errors caused by huge strings. In this case, the limit should have been updated as part of the patch that allowed user-specified names of arbitrary length. It's now changed to allow names up to 128 characters long, which seems reasonable considering the name will appear in the list of OpenVPN versions.

This is my fault for testing only in debug mode, where the checks are bypassed. I've fixed this in the source; the fix will be in the next beta release, probably early next week. I tested the fix works in normal mode. I apologize for making this process so painful and thank you for persevering and reporting the problems you've had.

I'm not sure there is such a thing as "too many security checks", although they do make it difficult to make changes.
